### ISSUE TITLE
Cannot fund accounts using wallet

### ISSUE THREAD

**Issue by TheLastGeneral** [jpmorganchase/cakeshop#84]
#### Description of problem
A generic "Sorry, transaction did not complete. Please try again." message appears when using the "Fund Accounts" widget in the Wallet section of CakeShop. The same error occurs without regard to which accounts are selected, whether or not they have sufficient funds, or what amounts (whole or fractional) are entered.

#### Expected behavior
The specified amount will be transferred from the source (from account) to the target (to account) as long as sufficient funds exist in the source.

#### Actual behavior
The error message "Sorry, transaction did not complete. Please try again." appears and funds are not transferred.

#### Steps to reproduce the behavior
After everything is up and running, open and login to CakeShop, click on the wallet section, and select any valid accounts in the "Fund Accounts" widget. Ensure that the "From Account" has sufficient funds to cover the amount being transferred. Enter the amount and click the "Transfer" button.

#### Description of investigation steps 
When attempting to send a value of 1 the logs showed that the jsonrpc payload being sent to Geth had the amount encoded as an raw whole number and there was an error indicating that non-string fields can not be unmarshaled.
```
[INFO ] 2018-08-25 13:09:25.452 [qtp166536076-22] GethHttpServiceImpl - Geth call: eth_sendTransaction
[ERROR] 2018-08-25 13:09:25.458 [qtp166536076-22] GethHttpServiceImpl - RPC request for {"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"from":"0x0638e1574728b6d862dd5d3a3e0942c3be47d996","to":"0x2e219248f44546d966808cdd20cb6c36df6efa82","value":1}],"id":"0x42"} failed with RPC request failed: invalid argument 0: json: cannot unmarshal non-string into Go struct field SendTxArgs.value of type *hexutil.Big
[WARN ] 2018-08-25 13:09:25.459 [qtp166536076-22] BaseController - Caught exception: RPC request failed: invalid argument 0: json: cannot unmarshal non-string into Go struct field SendTxArgs.value of type *hexutil.Big
```
When attempting to send a value of 10.01 the logs showed that the amount was encoded as a String, but there was an error indicating that only Long values are supported.
```
[WARN ] 2018-08-25 13:07:19.139 [qtp166536076-16] BaseController - Caught exception: JSON parse error: Cannot deserialize value of type `java.lang.Long` from String "10.01": not a valid Long value; nested exception is com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.lang.Long` from String "10.01": not a valid Long value
 at [Source: (PushbackInputStream); line: 1, column: 150] (through reference chain: com.jpmorgan.cakeshop.model.json.WalletPostJsonRequest["newBalance"])
org.springframework.http.converter.HttpMessageNotReadableException: JSON parse error: Cannot deserialize value of type `java.lang.Long` from String "10.01": not a valid Long value; nested exception is com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.lang.Long` from String "10.01": not a valid Long value
```

#### Description of unsuccessful resolution attempt
I verified that the the data transfer object ( WalletPostJsonRequest ) from the browser to the the service ( WalletServiceImpl ) did indeed only support Long values, and therefore modified those objects to serialize the Long value as a String prior to sending the value to Geth (via GethHttpService). While the jsonrpc payload now contained a valid Long value properly quoted as a String, the RPC requests still failed.

#### Description of successful potential resolution
Checking https://github.com/ethereum/wiki/wiki/JSON-RPC#hex-value-encoding indicates that RPC values must be hex encoded, and com.jpmorgan.cakeshop.util.AbiUtils contains a helper method to convert Long values to Hex Strings, so I updated com.jpmorgan.cakeshop.service.impl.WalletServiceImpl to Hex encode the amount value prior to sending. After implementing his change, positive whole numbers within the valid java long range worked successfully. The successful modification can be observed at: https://github.com/TheLastGeneral/cakeshop/commit/cdac818859c8158b9b203b20e5f6813546fd669e

#### Further Questions
What units are the amounts that are being transferred? 

I am led to believe that the amounts are ETH because the account balances section of the wallet shows balances in ETH and when transferring an amount of 1 the target account's balances get incremented by a full 1 ETH. However, the balances are displayed to two decimal places and there is no way to transfer fractional amounts because the amount is stored and transferred as a Long. If the amount is actually Wei, and the ETH label is incorrect, then a Long will not be of sufficient size to hold much value. If not Wei, and actually ETH, then there is no way to transfer small amounts using this wallet.

#### System information

Cakeshop version or commit hash: Cakeshop 0.10.0 Build 5c97592e

Cakeshop runtime configuration: stand alone 

Geth/Quorum version testing against: v1.7.2-stable/darwin-amd64/go1.10.2

Cakeshop OS & Version: OSX 10.10.5

Browser OS & Version: Firefox 61.0.2

**Comment by TheLastGeneral** [comment]
Followup on the question regarding "What units are the amounts that are being transferred?":

Checking https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getbalance indicates that account balances are reported in Wei. However, since a Java Long is being used internally in CakeShop to represent the amount being transferred, and it is being used as a signed number, then the maximum amount that can be transferred at one time is theoretically (2^63)-1 = 9223372036854775807. 

9223372036854775807 Ether == 9.223372036854775807e+36 Wei
9223372036854775807 Wei ==  9.223372036854775807 Ether

Regardless of whether the units are Ether or Wei, in either case larger numbers will need to be used because ETH account balances are typically represented by unsigned 256 bit values.

#### Verified problem exists
To verify whether or not this is actually a problem I attempted to transfer an amount of 9223372036854775808 and received a deserialization error:
`[WARN ] 2018-08-26 12:05:25.930 [qtp194415222-22] BaseController - Caught exception: JSON parse error: Cannot deserialize value of type `java.lang.Long` from String "9223372036854775808": not a valid Long value; nested exception is com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.lang.Long` from String "9223372036854775808": not a valid Long value
 at [Source: (PushbackInputStream); line: 1, column: 150] (through reference chain: com.jpmorgan.cakeshop.model.json.WalletPostJsonRequest["newBalance"])
org.springframework.http.converter.HttpMessageNotReadableException: JSON parse error: Cannot deserialize value of type `java.lang.Long` from String "9223372036854775808": not a valid Long value; nested exception is com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.lang.Long` from String "9223372036854775808": not a valid Long value
 at [Source: (PushbackInputStream); line: 1, column: 150] (through reference chain: com.jpmorgan.cakeshop.model.json.WalletPostJsonRequest["newBalance"])
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.readJavaType(AbstractJackson2HttpMessageConverter.java:241) ~[spring-web-5.0.7.RELEASE.jar!/:5.0.7.RELEASE]`

#### Converting Longs to BigIntegers
I modified  com.jpmorgan.cakeshop.model.json.WalletPostJsonRequest to use BigIntegers instead of Longs for the wallet balances, and since the com.jpmorgan.cakeshop.util.AbiUtils has toHex methods that support both Longs and BigIntegers, no additional changes were required to use the previous fix made to com.jpmorgan.cakeshop.service.impl.WalletServiceImpl to send amount values using hex encoding. While this worked for some values, it brought to light a compatability issue with hex encoding requirements ( https://github.com/ethereum/wiki/wiki/JSON-RPC#hex-value-encoding ) that no leading zeroes are allowed for non-zero values: 

> WRONG: 0x0400 (no leading zeroes allowed)


`
ERROR] 2018-08-26 12:16:56.039 [qtp638695843-21] GethHttpServiceImpl - RPC request for {"jsonrpc":"2.0","method":"eth_sendTransaction","params":[{"from":"0x0638e1574728b6d862dd5d3a3e0942c3be47d996","to":"0x2e219248f44546d966808cdd20cb6c36df6efa82","value":"0x008000000000000001"}],"id":"0x42"} failed with RPC request failed: invalid argument 0: json: cannot unmarshal hex number with leading zero digits into Go struct field SendTxArgs.value of type *hexutil.Big
[WARN ] 2018-08-26 12:16:56.049 [qtp638695843-21] BaseController - Caught exception: RPC request failed: invalid argument 0: json: cannot unmarshal hex number with leading zero digits into Go struct field SendTxArgs.value of type `*hexutil.Big`
 
#### Removing leading zeros from hex encoded values
It turns out that the Java toHexString method of a Long does not produce leading zero's so the AbiUtils.toHex(Long) method worked well, but AbiUtils.toHex(BigInteger) actually uses BouncyCastle's hex encoder, which does produce leading zeros. This is likely to be a compatibility issue that should be reviewed, however, in order to not introduce unintended consequences in case anyone was depending on having leading zeros, I created a new method in AbiUtils for explicitly converting BigIntegers  to Hex Strings without leading zeros.https://github.com/TheLastGeneral/cakeshop/commit/eb013a4a450356017c420abc0b7435162a331886 With this update, large (and small and zero value) whole number amounts were successfully transferred.

**Comment by TheLastGeneral** [comment]
Opened pull #85 based upon the current JPMC quorum-ui branch instead of @yourcodesucks updated  fork, so just the minimal related changes need to be reviewed at this time.  These and other changes have also already been accepted into yourcodesucks quorum-ui branch https://github.com/yourcodesucks/cakeshop/commit/072cb3ff9d418c5e6ac6535eb84b07d7b2db5a1d

**Comment by bmcd87** [comment]
Fixed in #95
